runtime/compiler/trj9/optimizer/ValueJProfiling.cpp

```diff
@@ -0,0 +1,883 @@
+ * \param base Base address to load from.
+ * \param index Optional index.
+ * \param offset Optional offset from the base.
+ */
+TR::Node *
+TR_ValueJProfiling::effectiveAddress(TR::DataType dataType, TR::Node *base, TR::Node *index, TR::Node *offset)
+   {
+   if (offset)
+      {
+      if (offset->getDataType() == TR::Int64)
+         base = TR::Node::create(base, TR::aladd, 2, base, offset);
+      else if (offset->getDataType() == TR::Int32)
+         base = TR::Node::create(base, TR::aiadd, 2, base, offset);
+      else
+         TR_ASSERT_FATAL(0, "Invalid type for address calculation integer");
+      }
+
+   if (index)
+      {
+      uint8_t size = TR::DataType::getSize(dataType);
+      if (index->getDataType() == TR::Int64)
+         base = TR::Node::create(base, TR::aladd, 2, base, TR::Node::create(base, TR::lmul, 2, index, TR::Node::lconst(base, size)));
+      else if (index->getDataType() == TR::Int32)
+         base = TR::Node::create(base, TR::aiadd, 2, base, TR::Node::create(base, TR::imul, 2, index, TR::Node::iconst(base, size)));
+      else
+         TR_ASSERT_FATAL(0, "Invalid type for address calculation integer");
+      }
+
+   return base;
+   }
+
+/*
+ * Given an address in X86's LEA form, generate a node structure to load a value from the memory address of the desired
+ * data type. Address would be: base + width(dataType) * index + offset
+ *
+ * \param dataType Date type of value to load. Its width is used for calculations if index is specified.
+ * \param base Base address to load from.
+ * \param index Optional index.
+ * \param offset Optional offset from the base.
+ */
+TR::Node *
+TR_ValueJProfiling::loadValue(TR::Compilation *comp, TR::DataType dataType, TR::Node *base, TR::Node *index, TR::Node *offset)
+   {
+   base = effectiveAddress(dataType, base, index, offset);
+   TR::SymbolReference *symRef = comp->getSymRefTab()->findOrCreateArrayShadowSymbolRef(dataType, NULL);
+   TR::Node *result = TR::Node::createWithSymRef(base, indirectLoad(dataType), 1, base, symRef);
+
+   return result;
+   }
+
+/*
+ * Generate the helper call tree, which adds values to the table along the slowest path.
+ * Will return a treetop, with the call as its child.
+ *
+ * \param value Node representing the value to add to the table.
+ * \param table Address of the table's base.
+ */
+TR::Node *
+TR_ValueJProfiling::createHelperCall(TR::Compilation *comp, TR::Node *value, TR::Node *table)
+   {
+   TR::SymbolReference *profiler;
+   if (value->getDataType() == TR::Int32)
+      profiler = comp->getSymRefTab()->findOrCreateRuntimeHelper(TR_jProfile32BitValue, false, false, true);
+   else
+      profiler = comp->getSymRefTab()->findOrCreateRuntimeHelper(TR_jProfile64BitValue, false, false, true);
+
+#if defined(TR_HOST_POWER) || defined(TR_HOST_ARM)
+   profiler->getSymbol()->castToMethodSymbol()->setLinkage(TR_System);
+#else
```

Comment:
this seems ugly can we avoid the nested case here and make this a more clean if elif else?